### PR TITLE
Avoid hitting slurm limits for maximum job count / array size

### LINKED ISSUE #104: Avoid hitting slurm limits for maximum job count / array size
When trying to schedule too many tasks, slurm's config can make the submission fail. To avoid the limits, the tasks can be chunked accordingly.

See https://github.com/scalableminds/cluster_tools/pull/103 for a reference of implementation.

### PR DESCRIPTION
**Pull request by daniel-wer**
Honor the slurm limit for maximum array job size, by splitting up larger job batches into multiple smaller batches.
Honor the slurm limit for the maximum number of jobs which are allowed to be submitted by a user at the same time by looking up the number of currently submitted jobs and only submitting new batches if they fit within the limit.
Thanks to the great slurm testing setup, it was fairly easy to add tests for the new behavior :tada: 

Also, because I stumbled upon an error which should've been caught by the linter, I changed the linting config from a very conservative one to the one we're using for voxelytics and fixed the resulting linting errors.

For the review, "Hide Whitespace" is beneficial :)

- fixes https://github.com/scalableminds/cluster_tools/issues/104

### CODE REVIEW

File: cluster_tools/schedulers/slurm.py [hunk 1]
```diff
@@ -100,22 +160,61 @@ def inner_submit(
...
+                    "srun {} {}".format(cmdline, job_index_start),
+                ]
+            )
+
+            job_id_futures.append(futures.Future())
+            scripts.append("\n".join(script_lines))
+            ranges.append((job_index_start, job_index_end + 1))
+
+        job_sizes = [end - start for start, end in ranges]
+        # Replace with asyncio.run when upgrading to python 3.7+
```
Comment (philippotto) [inline cluster_tools/schedulers/slurm.py:202]: Aren't we executing vx with 3.8 typically? Will run_until_complete still work?

File: cluster_tools/schedulers/slurm.py [hunk 2]
```diff
@@ -100,22 +160,61 @@ def inner_submit(
...
+        loop = asyncio.get_event_loop()
+        loop.run_until_complete(
+            self.submit_once_allowed(scripts, job_sizes, job_id_futures)
         )
 
-        return self.submit_text("\n".join(script_lines))
+        return job_id_futures, ranges
+
+    async def submit_once_allowed(self, scripts, job_sizes, futures):
```
Comment (philippotto) [inline cluster_tools/schedulers/slurm.py:210]: cool :+1: probably, the first time we use asyncio, but seems like a good fit here :)

Comment (philippotto) [reply cluster_tools/schedulers/slurm.py:210]: hm, on a second thought: won't `inner_submit` block until all job ids are available due to `run_until_complete`? then, I don't really see the point of returning futures. also, client code could only react to job results after all jobs were scheduled. I think, the code should submit the futures into the event loop without blocking. Maybe `asyncio.run_coroutine_threadsafe` (https://docs.python.org/3.5/library/asyncio-task.html#asyncio.run_coroutine_threadsafe) fits here?

File: cluster_tools/schedulers/slurm.py [hunk 3]
```diff
@@ -100,22 +160,61 @@ def inner_submit(
...
-            [
-                "#!/bin/sh",
-                "#SBATCH --output={}".format(log_path),
-                '#SBATCH --job-name "{}"'.format(job_name),
-                job_array_line,
-            ]
-            + job_resources_lines
-            + [*additional_setup_lines, "srun {}".format(cmdline)]
+        max_array_size = self.get_max_array_size()
+        max_submit_jobs = self.get_max_submit_jobs()
```
Comment (philippotto) [inline cluster_tools/schedulers/slurm.py:164]: I noticed in the tests, that the conig values can also be set to `-1`. I assume, slurm does not return the `-1` as a value if it's set that way (otherwise, the other tests would fail probably), but maybe it's a good idea to ensure `val > 0` in these getters? I could imagine that other slurm versions might yield `-1` if we are unlucky.

File: test.py [hunk 4]
```diff
@@ -372,6 +374,72 @@ def test_slurm_cfut_dir():
...
+            with executor:
+                futures = executor.map_to_futures(square, range(10))
+
+                result = [fut.result() for fut in futures]
+                assert result == [i ** 2 for i in range(10)]
+
+                job_ids = {fut.cluster_jobid for fut in futures}
+                # The 10 work packages should have been scheduled as 5 separate jobs,
+                # because the cluster_tools schedule at most 1/3 of MaxSubmitJobs at once.
+                assert len(job_ids) == 5
```
Comment (philippotto) [inline test.py:403]: cool test :+1:

File: test.py [hunk 5]
```diff
@@ -372,6 +374,72 @@ def test_slurm_cfut_dir():
...
+    executor = cluster_tools.get_executor("slurm", debug=True)
+
+    assert executor.get_number_of_submitted_jobs() == 0
+
+    with executor:
+        futures = executor.map_to_futures(sleep, [1] * number_of_jobs)
+        assert executor.get_number_of_submitted_jobs() == number_of_jobs
+        concurrent.futures.wait(futures)
+
+    assert executor.get_number_of_submitted_jobs() == 0
```
Comment (philippotto) [inline test.py:424]: I think, the assertion should go into the with-block, as the with-block itself is also waiting for the executor to be finished. To be stricter, doing the assertion after the first waiting would be better imo.

**Review by philippotto — CHANGES_REQUESTED** [review]
Exquisite stuff :1st_place_medal: Only left a couple of questions and nitpicks :)

**Review by philippotto — APPROVED** [review]
Very cool! Also excellent tests :+1: I think, I found one last edge case, but other than that, everything looks brilliant :100:

File: cluster_tools/schedulers/slurm.py [hunk 6]
```diff
@@ -194,28 +205,19 @@ def inner_submit(
...
             scripts.append("\n".join(script_lines))
             ranges.append((job_index_start, job_index_end + 1))
 
         job_sizes = [end - start for start, end in ranges]
-        # Replace with asyncio.run when upgrading to python 3.7+
-        loop = asyncio.get_event_loop()
-        loop.run_until_complete(
-            self.submit_once_allowed(scripts, job_sizes, job_id_futures)
+
+        self.submit_thread = JobSubmitThread(
```
[the remark this answers is not in the document]
Comment (daniel-wer) [reply cluster_tools/schedulers/slurm.py]: Very good call. I implemented your suggestion and also extended the tests respectively.

**Comment by philippotto** [comment]
:+1: changes look good now and the CI also seems to be happy :)

**Comment by daniel-wer** [comment]
Thanks for having another look :)
I'll test this in a real world scenario today and will merge if everything works as expected.
